compendium/ConfigurationAdmin/src/ConfigurationAdminImpl.cpp

```diff
@@ -684,21 +684,24 @@ ConfigurationAdminImpl::AddingService(
   const ServiceReference<cppmicroservices::service::cm::ManagedService>&
     reference)
 {
-  // Lock the configurations repository so no configuration objects can be
-  // added or removed while AddingService is processing the new service.
-  std::lock_guard<std::mutex> lk{ configurationsMutex };
- 
-  const auto pid = getPidFromServiceReference(reference);
-  if (pid.empty()) {
-    const auto bundle = reference.GetBundle();
-    const auto serviceID = std::to_string(cppmicroservices::any_cast<long>(
-      reference.GetProperty(Constants::SERVICE_ID)));
-    logger->Log(
-      SeverityLevel::LOG_WARNING,
-      "Ignoring ManagedService with ID " + serviceID + " from bundle " +
-        (bundle ? bundle.GetSymbolicName() : "Unknown") +
-        " as it does not have a service.pid or component.name property");
-    return nullptr;
+  std::string pid;
+  {
+    // Lock the configurations repository so no configuration objects can be
+    // added or removed while AddingService is processing the new service.
+    std::lock_guard<std::mutex> lk{ configurationsMutex };
+
+    pid = getPidFromServiceReference(reference);
+    if (pid.empty()) {
+      const auto bundle = reference.GetBundle();
+      const auto serviceID = std::to_string(cppmicroservices::any_cast<long>(
+        reference.GetProperty(Constants::SERVICE_ID)));
+      logger->Log(
+        SeverityLevel::LOG_WARNING,
+        "Ignoring ManagedService with ID " + serviceID + " from bundle " +
+          (bundle ? bundle.GetSymbolicName() : "Unknown") +
+          " as it does not have a service.pid or component.name property");
+      return nullptr;
+    }
   }
   auto managedService = cmContext.GetService(reference);
```

Comment:
I think releasing the lock here may result in the double Updated race condition happening again. If another thread creates and updates the configuration object while the GetService is happening, then an Updated notification will be sent for that update and the AddingService code will send another one at the end of the AddingService method (line 751)  To avoid the second Updated notification, AddingService can only send notifications for configuration objects that were created and updated before the AddingService method was called. One solution might be to move the check about whether or not the configuration object exists and has been updated  into the locked code before the lock is released for the GetService call.

Reply:
FYI, after locally modifying the `ConfigAdminTests.testDuplicateUpdated ` test to introduce multiple threads and running in a loop, I was able to get a double update on the config object:

```
Note: Google Test filter = ConfigAdminTests.testDuplicateUpdated
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ConfigAdminTests
[ RUN      ] ConfigAdminTests.testDuplicateUpdated
D:\Dev\CppMicroServices\cppms-ds-deadlock\CppMicroServices\compendium\ConfigurationAdmin\test\TestConfigAdmin.cpp(553): error: Expected equality of these values:
  serviceFactory->getUpdatedCounter("cm.testfactory~0")
    Which is: 2
  1
```
so this change definitely regresses the fix made in commit 664f5f93479c31f4c2d4592145002ac54fe69a67

Reply:
side note;

after removing the scoping of the lock in the `AddingService` ServiceTracker callbacks (essentially reverting back to the same state as 664f5f93479c31f4c2d4592145002ac54fe69a67), I get this failure after running many iterations of the test:

```
Repeating all tests (iteration 340042) . . .

Note: Google Test filter = ConfigAdminTests.testDuplicateUpdated
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ConfigAdminTests
[ RUN      ] ConfigAdminTests.testDuplicateUpdated
D:\Dev\CppMicroServices\cppms-ds-deadlock\CppMicroServices\compendium\ConfigurationAdmin\test\TestConfigAdmin.cpp(553): error: Expected equality of these values:
  serviceFactory->getUpdatedCounter("cm.testfactory~0")
    Which is: 0
  1
```

Reply:
to confirm, there is an pre-existing race condition that causes the config update to be missed which manifest when running ComfigAdminTests.testConcurrentDuplicateUpdated in a loop and observing the following test failure:

```
[ RUN      ] ConfigAdminTests.testDuplicateUpdated
D:\Dev\CppMicroServices\cppms-ds-deadlock\CppMicroServices\compendium\ConfigurationAdmin\test\TestConfigAdmin.cpp(553): error: Expected equality of these values:
  serviceFactory->getUpdatedCounter("cm.testfactory~0")
    Which is: 0
  1
```

This race happens when the configuration object has not been created/updated and the service tracker adding the ManagedServiceFactory service has not yet added the TrackedServiceWrapper object to the internal tracking map.  The `Configuration::UpdateIfDifferent` gets all the tracked ManagedServiceFactory objects and if it hasn't been added yet AND the ServiceTracker::AddingService completes before the configuration object is created, then the config update is missed completely.

Reply:
opened issue #716 to track this.